[1.21.3] Forge CoreMods cleanup

Jonathing opened the pull request:
This PR aims to clean up and optimize the existing CoreMods found within Forge.

- This work was originally started in #10156 but has since been moved to this PR.
- Requires #10156 to be merged first.
- Fixes #10157.
- This PR will not be backported.

Jonathing committed:
Don't remap finalizeSpawn in method_redirector

Jonathing committed:
method_redirector CoreMod code cleanup, ES6 feature usage

Jonathing committed:
Re-organize method_redirector CoreMod code

Jonathing committed:
Remove redundant code comment

Jonathing committed:
Turn ASMAPI reference into a const

Jonathing committed:
Made the revamped coremod significantly less stinky

Jonathing committed:
Added some debug logging to method_redirector

Jonathing committed:
Print an error if no redirections were applied to a target class

Jonathing committed:
Clarify in logs that this is a Forge bug

Jonathing committed:
Some additional improvements moved over from old PR

Jonathing committed:
Apply the correct descriptor for method redirect

Jonathing committed:
Fix additional errors I didn't catch

Jonathing committed:
Check class replacements list (this is a stopgap fix for now)

Jonathing committed:
PROGRESS: Only search targeted methods in classes

Jonathing committed:
Clean it up. It fucking works!!!! :) :) :)

Jonathing commented:
The new `method_redirector` CoreMod has been tested and works just as well as it did before. I've modified `finalize_spawn_targets.json` to declare both the class name and method names where we need to redirect method calls. This gets rid of the need for the CoreMod to search the entire class.

Jonathing committed:
Mention the replacement name if we're erroring

Jonathing committed:
Add license headers

Jonathing committed:
Fixed and updated findFinalizeSpawnTargets

Jonathing committed:
Consistent sorting in the json generator

PaintNinja commented on src/main/resources/coremods/method_redirector.js:
```diff
@@ -1,77 +1,98 @@
+        for (let methodString of getClassTargetMethods(clazz, replacement)) {
+            // cut up the string since we've put both the name and desc as the target
+            const methodName = methodString.substring(0, methodString.indexOf('('));
+            const methodDesc = methodString.substring(methodString.indexOf('('));
```
```suggestion
            // cut up the string since we've put both the name and desc as the target
            const splitPos = methodString.indexOf('(');
            const methodName = methodString.substring(0, splitPos);
            const methodDesc = methodString.substring(splitPos);
```

Jonathing committed:
Don't split the methodString twice

Applied through GitHub.com with https://github.com/MinecraftForge/MinecraftForge/pull/10159#discussion_r1896724095

Co-authored-by: Paint_Ninja <PaintNinja@users.noreply.github.com>

Jonathing committed:
Check all synthetic methods of a class

Jonathing committed:
Don't add synthetic methods to targets json (still add class)
